ChiModules/LinearBoltzmannSolver/Acceleration/diffusion_mip_02a_assembleAand_b.cc

```diff
@@ -0,0 +1,377 @@
+#include "diffusion_mip.h"
+
+#include "ChiMesh/MeshContinuum/chi_meshcontinuum.h"
+
+#include "ChiMath/SpatialDiscretization/spatial_discretization.h"
+#include "ChiMath/SpatialDiscretization/FiniteElement/finite_element.h"
+
+#include "ChiPhysics/PhysicsMaterial/transportxsections/material_property_transportxsections.h"
+
+#include "chi_runtime.h"
+#include "chi_log.h"
+#include "ChiTimer/chi_timer.h"
+#include "ChiConsole/chi_console.h"
+
+#define DefaultBCDirichlet BoundaryCondition{BCType::DIRICHLET,{0,0,0}}
+
+//###################################################################
+/**Assembles both the matrix and the RHS using quadrature points. These
+ * routines exist for implementing MMS.*/
+void lbs::acceleration::DiffusionMIPSolver::
+  AssembleAand_b_wQpoints(const std::vector<double>& q_vector)
+{
+  const std::string fname = "lbs::acceleration::DiffusionMIPSolver::"
+                            "AssembleAand_b_wQpoints";
+  if (m_A == nullptr or m_rhs == nullptr or m_ksp == nullptr)
+    throw std::logic_error(fname + ": Some or all PETSc elements are null. "
+                                   "Check that Initialize has been called.");
+  if (options.verbose)
+    chi::log.Log() << chi::program_timer.GetTimeString() << " Starting assembly";
+
+  lua_State* L = chi::console.consoleState;
+  const auto& source_function = options.source_lua_function;
+  const auto& solution_function = options.ref_solution_lua_function;
+
+  const size_t num_groups   = m_uk_man.unknowns.front().num_components;
+
+  VecSet(m_rhs, 0.0);
+
+  for (const auto& cell : m_grid.local_cells)
+  {
+    const size_t num_faces    = cell.faces.size();
+    const auto&  cell_mapping = m_sdm.GetCellMapping(cell);
+    const size_t num_nodes    = cell_mapping.NumNodes();
+    const auto   cc_nodes     = cell_mapping.GetNodeLocations();
+    const auto   qp_data      = cell_mapping.MakeVolumeQuadraturePointData();
+
+    const auto& xs = m_map_mat_id_2_xs.at(cell.material_id);
+
+    //=========================================== For component/group
+    for (size_t g=0; g<num_groups; ++g)
+    {
+      //==================================== Get coefficient and nodal src
+      const double Dg     = xs.Dg[g];
+      const double sigr_g = xs.sigR[g];
+
+      std::vector<double> qg(num_nodes, 0.0);
+      for (size_t j=0; j<num_nodes; j++)
+        qg[j] = q_vector[m_sdm.MapDOFLocal(cell, j, m_uk_man, 0, g)];
+
+      //==================================== Assemble continuous terms
+      for (size_t i=0; i<num_nodes; i++)
+      {
+        const int64_t imap = m_sdm.MapDOF(cell,i,m_uk_man,0,g);
+        double entry_rhs_i = 0.0;
```

Comment:
Put a comment as of why entry_rhs_i is set to 0 before the loop on j (even when things get accumulated in it through a loop on j). If you look, entry_aij is set to 0 inside the loop on j, not before it. IT feels fishy but I understand it is not

you are computing
rhi_i = int b_i(x) rhs(x)
so rhi_i = 0 is initialized at the correct location.
However, it happens that rhs(x) = sum_j q_j b_j(x) when source_function.empty() is true.

When source_function.empty() is FALSE, however, what you are doing is 
rhi_i = int b_i(x) rhs(x) = rhi_i = int b_i(x) \times (1) \times rhs(x)
and then you replace 1 by sum_j b_j(x) which is 1 by virtue of partitition of unity (sum of basis functions = 1)

Reply:
Mmmm. Good find. Corrected the code and set the comment.